mlir/lib/Gradient/Transforms/vector_lowering.cpp

```diff
@@ -0,0 +1,229 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Bufferization/IR/Bufferization.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Gradient/IR/GradientOps.h"
+#include "Gradient/Transforms/Patterns.h"
+
+using namespace mlir;
+
+namespace catalyst {
+namespace gradient {
+/**
+ * A utility builder that aids in lowering dynamically-resizable parameter vectors.
+ *
+ * ParameterVectors are lowered to a size, a capacity, and a memref that stores the vector data.
+ * Each component is made mutable by being stored in a rank-0 memref, such that the vector data is
+ * lowered to (memref<memref<?x{element-type}>>, memref<index>, memref<index>).
+ */
+struct DynVectorBuilder {
+    Value dataField, sizeField, capacityField;
+    Type elementType;
+
+    static FailureOr<DynVectorBuilder> get(Location loc, TypeConverter *typeConverter,
+                                           TypedValue<ParameterVectorType> vector, OpBuilder &b)
+    {
+        SmallVector<Type> resultTypes;
+        if (failed(typeConverter->convertType(vector.getType(), resultTypes))) {
+            return failure();
+        }
+
+        auto unpacked = b.create<UnrealizedConversionCastOp>(loc, resultTypes, vector);
+        return DynVectorBuilder{.dataField = unpacked.getResult(0),
+                                .sizeField = unpacked.getResult(1),
+                                .capacityField = unpacked.getResult(2),
+                                .elementType = vector.getType().getElementType()};
+    }
+
+    FlatSymbolRefAttr getOrInsertPushFunction(Location loc, ModuleOp moduleOp, OpBuilder &b) const
+    {
+        MLIRContext *ctx = b.getContext();
+        std::string funcName = "__grad_vec_push";
+        llvm::raw_string_ostream nameStream{funcName};
+        nameStream << elementType;
+        if (moduleOp.lookupSymbol<func::FuncOp>(funcName))
+            return SymbolRefAttr::get(ctx, funcName);
```

Comment:
This is also not recommended by `clang-tidy` associated with `readability-braces-around-blocks`. Let's update this and other places as follows,
```suggestion
        if (moduleOp.lookupSymbol<func::FuncOp>(funcName)) {
            return SymbolRefAttr::get(ctx, funcName);
        }
```